File: src/test/java/racingCar/Car.java
package racingCar;

public class Car {

    private static final int DEFAULT_POSITION = 0;

    private String name;
    private int position;

    public Car(String name) {
        this(name, DEFAULT_POSITION);
    }

    public Car(String name, int position) {
        validateName(name);
        this.name = name;
        this.position = position;
    }

    public void move(MovingStrategy movingStrategy) {
        if (movingStrategy.movable()) {
            this.position++;
        }
    }

    private void validateName(String name) {
        if (name == null || name.isBlank() || name.length() > 5) {
            throw new IllegalArgumentException();
        }
    }

    public int getPosition() {
        return position;
    }
}

File: src/test/java/racingcar/Car.kt
package racingcar

data class Car (
    val name: String,
    var position: Int
) {
    init {
        require(name.length <= 5)
        {
            throw IllegalArgumentException("자동차 이름은 5글자를 넘을 수 없습니다.")
        }
    }

    fun move(condition: MovingStrategy) {
        if (condition.isMovable()) {
            position++
        }
    }

    companion object {
        fun init(name: String): Car {
            return Car(name, 0)
        }
    }
}